cdc/kv/shared_client.go

```diff
@@ -0,0 +1,1196 @@
+		return nil, errors.Trace(err)
+	}
+	return rpcCtx, nil
+}
+
+func (s *SharedClient) requestStore(
+	ctx context.Context, g *errgroup.Group,
+	storeID uint64, storeAddr string,
+) *requestedStore {
+	var rs *requestedStore
+	if rs = s.requestedStores[storeAddr]; rs != nil {
+		return rs
+	}
+
+	rs = &requestedStore{storeID: storeID, storeAddr: storeAddr}
+	s.requestedStores[storeAddr] = rs
+	for i := uint(0); i < s.config.GrpcStreamConcurrent; i++ {
+		stream := s.newStream(ctx, g, rs)
+		rs.streams = append(rs.streams, stream)
+	}
+
+	return rs
+}
+
+func (s *SharedClient) createRegionRequest(sri singleRegionInfo) *cdcpb.ChangeDataRequest {
+	return &cdcpb.ChangeDataRequest{
+		Header:       &cdcpb.Header{ClusterId: s.clusterID, TicdcVersion: version.ReleaseSemver()},
+		RegionId:     sri.verID.GetID(),
+		RequestId:    uint64(sri.requestedTable.subscriptionID),
+		RegionEpoch:  sri.rpcCtx.Meta.RegionEpoch,
+		CheckpointTs: sri.resolvedTs(),
+		StartKey:     sri.span.StartKey,
+		EndKey:       sri.span.EndKey,
+		ExtraOp:      kvrpcpb.ExtraOp_ReadOldValue,
+		FilterLoop:   s.filterLoop,
+	}
+}
+
+func (s *SharedClient) sendToStream(ctx context.Context, rs *requestedStore, stream *requestedStream) (err error) {
+	defer func() {
+		closeSendErr := stream.client.CloseSend()
+		log.Info("event feed close send for grpc stream",
+			zap.String("namespace", s.changefeed.Namespace),
+			zap.String("changefeed", s.changefeed.ID),
+			zap.Uint64("storeID", rs.storeID),
+			zap.String("addr", rs.storeAddr),
+			zap.Uint64("streamID", stream.streamID),
+			zap.String("reason", err.Error()),
+			zap.Error(closeSendErr))
+	}()
+
+	doSend := func(req *cdcpb.ChangeDataRequest) error {
+		if err := stream.client.Send(req); err != nil {
+			log.Warn("event feed send request to grpc stream failed",
+				zap.String("namespace", s.changefeed.Namespace),
+				zap.String("changefeed", s.changefeed.ID),
+				zap.Any("subscriptionID", SubscriptionID(req.RequestId)),
+				zap.Uint64("regionID", req.RegionId),
+				zap.Uint64("storeID", rs.storeID),
+				zap.String("addr", rs.storeAddr),
+				zap.Uint64("streamID", stream.streamID),
+				zap.Error(err))
+			return errors.Trace(err)
+		}
+		log.Debug("event feed send request to grpc stream success",
+			zap.String("namespace", s.changefeed.Namespace),
+			zap.String("changefeed", s.changefeed.ID),
+			zap.Any("subscriptionID", SubscriptionID(req.RequestId)),
+			zap.Uint64("regionID", req.RegionId),
+			zap.Uint64("storeID", rs.storeID),
+			zap.String("addr", rs.storeAddr),
+			zap.Uint64("streamID", stream.streamID))
+		return nil
+	}
+
+	fetchMoreReq := func() (sri singleRegionInfo, _ error) {
+		waitReqTicker := time.NewTicker(60 * time.Second)
+		defer waitReqTicker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return sri, ctx.Err()
+			case sri = <-stream.requests.Out():
+				return sri, nil
+			case <-waitReqTicker.C:
+				if stream.countStates() == 0 {
+					return sri, errors.New("closed as idle")
```

Comment:
Please add some comments here to explain that the stream will be reinitialized when the next request is received.